pkg/cmd/accept/exec.go

```diff
@@ -61,95 +64,133 @@ func (o *Options) run() error {
 	return o.runWithClient(kubeClient, clusterClient)
 }
 
-func (o *Options) runWithClient(kubeClient *kubernetes.Clientset, clusterClient *clusterclientset.Clientset) error {
+func (o *Options) runWithClient(kubeClient *kubernetes.Clientset, clusterClient *clusterclientset.Clientset) (err error) {
 	for _, clusterName := range o.values.clusters {
-
-		csrs, err := kubeClient.CertificatesV1().CertificateSigningRequests().List(context.TODO(),
-			metav1.ListOptions{
-				LabelSelector: fmt.Sprintf("%v = %v", clusterLabel, clusterName),
+		if o.wait == 0 {
+			_, err = o.accept(kubeClient, clusterClient, clusterName)
+		} else {
+			err = wait.PollImmediate(1*time.Second, time.Duration(o.wait)*time.Second, func() (bool, error) {
+				return o.accept(kubeClient, clusterClient, clusterName)
 			})
+		}
 		if err != nil {
 			return err
 		}
-		var csr *certificatesv1.CertificateSigningRequest
-		for _, item := range csrs.Items {
-			//Does not have the correct name prefix
-			if !strings.HasPrefix(item.Spec.Username, userNameSignaturePrefix) {
-				continue
-			}
-			//Check groups
-			var group string
-			for _, g := range item.Spec.Groups {
-				if g == groupName {
-					group = g
-					break
-				}
-			}
-			//Does not contain the correct group
-			if len(group) == 0 {
-				continue
+	}
+	return nil
+}
+
+func (o *Options) accept(kubeClient *kubernetes.Clientset, clusterClient *clusterclientset.Clientset, clusterName string) (bool, error) {
+	csrApproved, err := o.approveCSR(kubeClient, clusterName)
+	if err != nil {
+		return false, err
+	}
+	mcUpdated, err := o.updateManagedCluster(clusterClient, clusterName)
+	if err != nil {
+		return false, err
+	}
+	if csrApproved && mcUpdated {
+		return true, nil
+	}
+	return false, nil
+
+}
+
+func (o *Options) approveCSR(kubeClient *kubernetes.Clientset, clusterName string) (bool, error) {
+	csrs, err := kubeClient.CertificatesV1().CertificateSigningRequests().List(context.TODO(),
+		metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%v = %v", clusterLabel, clusterName),
+		})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	var csr *certificatesv1.CertificateSigningRequest
+	for _, item := range csrs.Items {
+		//Does not have the correct name prefix
+		if !strings.HasPrefix(item.Spec.Username, userNameSignaturePrefix) {
+			continue
+		}
+		//Check groups
+		var group string
+		for _, g := range item.Spec.Groups {
+			if g == groupName {
+				group = g
+				break
 			}
-			//Check if already approved or denied
-			done := false
-			for _, c := range item.Status.Conditions {
-				if c.Type == certificatesv1.CertificateApproved || c.Type == certificatesv1.CertificateDenied {
-					done = true
-					break
-				}
+		}
+		//Does not contain the correct group
+		if len(group) == 0 {
+			continue
+		}
+		//Check if already approved or denied
+		for _, c := range item.Status.Conditions {
```

Comment:
would you move this to a seperate IsCSRApproved func? like this
https://github.com/open-cluster-management-io/registration/blob/e83faae5f8b9646e0244f180839e38e186dcffc4/pkg/clientcert/certificate.go#L161

Reply:
yes can do

Reply:
but will be isApprovedOrDenied

Reply:
but the link... doesn't check if already denied... if already denied should we give a chence to approve it again?

Reply:
if it is already denied, it cannot be approve anymore

Reply:
then my code is good, because the example you gave still allow an approve after get denied,

Reply:
my point is move those part into a separate func